### PR TITLE
[compiler] Expose relay ops registered as torch.ops.tvm.__

### PR DESCRIPTION
**Pull request by bwasti**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44 [compiler] Expose `to_relay` API which hands the Relay graph over in Python
* **#43 [compiler] Expose relay ops registered as torch.ops.tvm.__**

### CODE REVIEW

**Comment by bwasti** [comment]
There is currently a bug lurking here regarding the lifetime management of DLManagedTensors.  TVM stores NDArrays in the graph runtime and on program exit will attempt to delete them, assuming that it has ownership of the DLManagedTensors' tensors.  This ends up breaking everything and the program segfaults.  A small patch to get around this is

```
diff --git a/src/runtime/ndarray.cc b/src/runtime/ndarray.cc
index 39c17b8b..92747b1b 100644
--- a/src/runtime/ndarray.cc
+++ b/src/runtime/ndarray.cc
@@ -190,8 +190,10 @@ void NDArray::CopyFromTo(DLTensor* from,
 using namespace tvm::runtime;

 void NDArrayDLPackDeleter(DLManagedTensor* tensor) {
-  static_cast<NDArray::Container*>(tensor->manager_ctx)->DecRef();
-  delete tensor;
+  if (tensor) {
+    static_cast<NDArray::Container*>(tensor->manager_ctx)->DecRef();
+    delete tensor;
+  }
 }

 int TVMArrayAlloc(const tvm_index_t* shape,
```

but realistically we should revisit the ownership model and how values are passed in to TVM

File: torch_tvm/operators.cpp [hunk 1]
```diff
@@ -14,12 +20,50 @@ std::unordered_map<Symbol, TVMOpFunctor>& getTVMOperatorMap() {
...
+        wrapper_graphs.emplace_back();
+        auto &wrapper_graph = wrapper_graphs.back();
+        std::vector<Value *> torch_inputs;
+        for (const auto &inp : schema.arguments()) {
+          torch_inputs.emplace_back(wrapper_graph.addInput());
+        }
+        Node* node =
+            wrapper_graph.create(op.sym, torch_inputs, schema.returns().size());
+        wrapper_graph.appendNode(node);
+        wrapper_graph.registerOutput(node->output());
```
Comment (wanchaol) [inline torch_tvm/operators.cpp]: this is a bit confusing. `wrapper_graph` create and append the node into its graph, then it does not ever get used when we creating the `SingletonSubgraph`?

File: torch_tvm/operators.cpp [hunk 2]
```diff
@@ -14,12 +20,50 @@ std::unordered_map<Symbol, TVMOpFunctor>& getTVMOperatorMap() {
...
+
+        Symbol tvm_sym = Symbol::fromQualString("tvm::CompilationGroup");
+        node = SubgraphUtils::createSingletonSubgraph(node, tvm_sym);
+        auto cc = std::make_shared<TVMCompiler>(node);
+
+        // NB: We assume all relay ops are pure
+        auto options = OperatorOptions().aliasAnalysis(AliasAnalysisKind::PURE);
+        // We don't know the true number of inputs so we make it varargs
```
Comment (wanchaol) [inline torch_tvm/operators.cpp]: we do know the true number of arguments right? even if in runtime, the number of inputs are the same as `schema.arguments()` I believe.

File: torch_tvm/compiler.cpp [hunk 3]
```diff
@@ -214,7 +216,7 @@ void TVMCompiler::run(Stack& stack) {
     // or fall back to the JIT interpreter for execution
     tvm::relay::Function tvm_func;
     try {
-      tvm_func = convertToRelay(subgraph_, &input_values_);
+      tvm_func = convertToRelay(subgraph_, &cache_[spec].input_values);
```
[the remark this answers is not in the document]
Comment (bwasti) [reply torch_tvm/compiler.cpp]: C++ unordered_map's [] creates an object if it doesn't exist

File: torch_tvm/operators.cpp (hunk 1 above)
Comment (bwasti) [reply torch_tvm/operators.cpp]: it's a lifetime management hack

File: torch_tvm/operators.cpp (hunk 2 above)
Comment (bwasti) [reply torch_tvm/operators.cpp]: oh yea we do! dated comment

**Review by wanchaol — APPROVED** [review]
looks good

**Pull request closed without merging**